### PR TITLE
switch expression and integral promotions

### PR DESCRIPTION
**Pull request by WalterBright**
Clarify what switch expressions are accepted, and when integral promotions on it are done. This documents existing behavior.

### CODE REVIEW

**Review by PetarKirov — APPROVED** [review]

**Comment by PetarKirov** [comment]
> WalterBright requested a review from andralex

I missed that. @andralex feel free to open a follow-up PR, if you want to change the wording.

File: spec/statement.dd [hunk 1]
```diff
@@ -1103,9 +1103,17 @@ $(GNAME StatementNoCaseNoDefault):
     $(GLINK ScopeBlockStatement)
 )
 
-        $(P $(EXPRESSION) is evaluated. The result type T must be
-        of integral type or $(CODE char[]), $(CODE wchar[]) or $(CODE dchar[]).
-        The result is
+        $(P $(I Expression) is evaluated.
+        If the type of the $(I Expression) is an enum,
```
Comment (andralex) [inline spec/statement.dd:1107]: The rule to follow here is, anything that is not an English word must have some annotation other than plain text. The word "enum" is not in the dictionary, so it needs to be in code font.

File: spec/statement.dd [hunk 2]
```diff
@@ -1103,9 +1103,17 @@ $(GNAME StatementNoCaseNoDefault):
...
+        it is repeatedly converted
+        to its $(GLINK2 enum, EnumBaseType) until it isn't. Then,
+        if the type is of integral type,
```
Comment (andralex) [inline spec/statement.dd:1110]: A type does not have a type. So "if the type is integral" would be an alternative.

File: spec/statement.dd [hunk 3]
```diff
@@ -1103,9 +1103,17 @@ $(GNAME StatementNoCaseNoDefault):
...
+        the $(I Expression) undergoes $(DDSUBLINK spec/type, integer-promotions, Integer Promotions).
+        Then, the type of the $(I Expression) must be either of integral type or
```
Comment (andralex) [inline spec/statement.dd:1112]: "the type of. [...] must be [...] of integral type" again does not work, so perhaps: "Then, the resulting type must be either integral or ..."

**Comment by andralex** [comment]
I've added a few notes, thanks.

**Comment by WalterBright** [comment]
@andralex see https://github.com/dlang/dlang.org/pull/3075

**Pull request merged**